### PR TITLE
gql/iriref : Parsing of IRI chars in gql.

### PR DESCRIPTION
**Pull request by ashishnegi**
Following the grammar :

```
IRIREF ::= '<' ([^#x00-#x20<>"{}|^`\] | UCHAR)* '>'
UCHAR ::= '\u' HEX HEX HEX HEX | '\U' HEX HEX HEX HEX HEX HEX HEX HEX
HEX ::= [0-9] | [A-F] | [a-f]
```

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/524)

### CODE REVIEW

File: gql/parser_test.go [hunk 1]
```diff
@@ -1298,6 +1298,45 @@ func TestParseGenerator(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestParseIRIRef(t *testing.T) {
+	query := `{
+		me(id:<http://helloworld.com/how/are/you>) {
```
Comment (ashishnegi) [inline gql/parser_test.go]: Is there any other place in query where `predicates` can come ?

**Comment by manishrjain** [comment]
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[gql/parser_test.go, line 1303 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDNOha_ahsOMvxakD--r1-1303:-KbDQCHOAOR_cUnJ1E5c:bacj428) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/parser_test.go#L1303)):*
<details><summary><i>Previously, ashishnegi (Ashish Negi) wrote…</i></summary><blockquote>

Is there any other place in query where `predicates` can come ?
</blockquote></details>

It could come in a function at the root as well.

```
me(anyof(<http://....>, "good better bad")) {
}
```

---

*[gql/parser_test.go, line 1305 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDQB9FtsyaqW7-3lWV:-KbDQB9FtsyaqW7-3lWW:bffn0gb) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/parser_test.go#L1305)):*
> ```Go
> 		me(id:<http://helloworld.com/how/are/you>) {
>                         <http://verygood.com/what/about/you>
> 			friends @filter(allof(<http://verygood.com/what/about/you>, "good better bad")){
> ```

Formatting issue.

---

*[gql/parser_test.go, line 1323 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDQ8hwaIBmwnzGaHZb:-KbDQ8hxu9zQbd6aSEcY:bffn0gb) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/parser_test.go#L1323)):*
> ```Go
> 	query := `{
> 		me(id:<http://helloworld.com/how/are/ you>) {
>                 }
> ```

Formatting issue.

---

*[gql/parser_test.go, line 1332 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDOi-Oa54KPZ-mJpLC:-KbDOi-Oa54KPZ-mJpLD:b-73v0v1) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/parser_test.go#L1332)):*
> ```Go
> func TestParseIRIRefInvalidChar(t *testing.T) {
> 	query := `{
> 		me(id:<http://helloworld.com/how/are/^you>) {
> ```

How about a space after `id:`?

---

*[gql/state.go, line 567 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDPw9eoAJP3yV3pjHK:-KbDPw9eoAJP3yV3pjHL:b-jdnqk3) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/state.go#L567)):*
> ```Go
> 			l.Backup()
> 			return false
> 		} else {
> ```

No else after return.

---

*[gql/state_test.go, line 145 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDQ2yEjKAJXznzoKKw:-KbDQ2yEjKAJXznzoKKx:bf52yxw) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/state_test.go#L145)):*
> ```Go
> 	input := `
> 	query testQuery {
> 		me(_xid_: <http://helloworld.com/how/are/you>) {
> ```

`_xid_:` is no longer a valid syntax.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/524)*

**Comment by ashwin95r** [comment]
Reviewed 3 of 3 files at r1, 3 of 6 files at r2.
Review status: 3 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[schema/state.go, line 447 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbE71B_DZ_L1zfQcjWK:-KbE71B_DZ_L1zfQcjWL:bpwwvz2) ([raw file](https://github.com/dgraph-io/dgraph/blob/e6119f712e0f2e144ce3f17310c4074454622397/schema/state.go#L447)):*
> ```Go
> 
> // IRIREF ::= '<' ([^#x00-#x20<>"{}|^`\] | UCHAR)* '>'
> func isIRIChar(r rune, l *lex.Lexer) bool {
> ```

Would be nice if we could move these functions to x package and then use it in gql, ref, schema packages. So in case we change the spec we'd have to modify it only at one place.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/524)*

**Comment by ashishnegi** [comment]
Review status: 3 of 6 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[gql/parser_test.go, line 1303 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDNOha_ahsOMvxakD--r1-1303:-KbE7FRpfWWVYruOmloK:bufpfon) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/parser_test.go#L1303)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

It could come in a function at the root as well.

```
me(anyof(<http://....>, "good better bad")) {
}
```
</blockquote></details>

Added test for it as well.

---

*[gql/parser_test.go, line 1305 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDQB9FtsyaqW7-3lWV:-KbE7EbHxN0CR86DqJti:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/parser_test.go#L1305)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Formatting issue.
</blockquote></details>

Done.

---

*[gql/parser_test.go, line 1323 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDQ8hwaIBmwnzGaHZb:-KbE87dky4JmjzyoNeyZ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/parser_test.go#L1323)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Formatting issue.
</blockquote></details>

Done.

---

*[gql/parser_test.go, line 1332 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDOi-Oa54KPZ-mJpLC:-KbDf-GGjeta7CfbAld6:buc95kq) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/parser_test.go#L1332)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

How about a space after `id:`?
</blockquote></details>

that will work too. :)
will put it. Done.

---

*[gql/state_test.go, line 145 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDQ2yEjKAJXznzoKKw:-KbDfAaMpOZgY-gcdRnw:bqyzbo8) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/state_test.go#L145)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

`_xid_:` is no longer a valid syntax.
</blockquote></details>

uh.. right . it think it does not matter in lexer.
will change.

---

*[schema/state.go, line 447 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbE71B_DZ_L1zfQcjWK:-KbE7weYpJCFaZhOzJLl:btx1q5s) ([raw file](https://github.com/dgraph-io/dgraph/blob/e6119f712e0f2e144ce3f17310c4074454622397/schema/state.go#L447)):*
<details><summary><i>Previously, ashwin95r (Ashwin Ramesh) wrote…</i></summary><blockquote>

Would be nice if we could move these functions to x package and then use it in gql, ref, schema packages. So in case we change the spec we'd have to modify it only at one place.
</blockquote></details>

can i do it as separate next PR ?
this PR already has schema and gql changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/524)*

**Comment by ashishnegi** [comment]
Review status: 3 of 6 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[gql/state.go, line 567 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbDPw9eoAJP3yV3pjHK:-KbE8IpwUvTkjXWleiBv:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9dd7ee8f418f1421220d0de37bfb3c0b280cdf09/gql/state.go#L567)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

No else after return.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/524)*

**Comment by ashwin95r** [comment]
:lgtm:

---

Review status: 3 of 6 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[schema/state.go, line 447 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-KbE71B_DZ_L1zfQcjWK:-KbEABXIn_f0NyN7Wx_3:b-80tlqk) ([raw file](https://github.com/dgraph-io/dgraph/blob/e6119f712e0f2e144ce3f17310c4074454622397/schema/state.go#L447)):*
<details><summary><i>Previously, ashishnegi (Ashish Negi) wrote…</i></summary><blockquote>

can i do it as separate next PR ?
this PR already has schema and gql changes.
</blockquote></details>

Sure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-:-KbEAHolgrLer3pcEoKc:bnfp4nl)*

**Comment by manishrjain** [comment]
:lgtm:

---

Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/524#-:-KbED61B1CxRBQW7yVn4:bnfp4nl)*

**Pull request closed without merging**